File: src/dsa/arrays/inserting/sorted_inserting_manually.ts
const sorted_arr = [10, 12, 23, 34, 45];
const new_element_3 = 25;
const new_sorted_arr = insertElementSorted(sorted_arr, new_element_3);

console.log(sorted_arr, 'Array antes da inserção');
console.log(new_sorted_arr, 'Array depois da inserção');

function insertElementSorted(arr: number[], element: number): number[] {
    const new_arr = new Array(arr.length + 1);
    let inserted = false;

    for (let i = 0, j = 0; i < arr.length; i++, j++) {
        if (inserted === false && arr[i] > element) {
            new_arr[j] = element;
            inserted = true;
            j++;
        }

        new_arr[j] = arr[i];
    }

    if (inserted === false) {
        new_arr[arr.length] = element;
    }

    return new_arr;
}

File: src/dsa/arrays/inserting/sorted_inserting_manually.rs
pub fn run() {
    let sorted_arr: Vec<i32> = vec![3, 6, 12, 23, 27, 32];
    let new_element: i32 = 18;

    println!("Array inicial: {:?}", sorted_arr);

    let new_sorted_arr: Vec<i32> = insert_sorted(sorted_arr, new_element);

    println!("Array final: {:?}", new_sorted_arr);
}

fn insert_sorted(arr: Vec<i32>, element: i32) -> Vec<i32> {
    let mut new_arr: Vec<i32> = Vec::with_capacity(arr.len() + 1);
    let mut inserted: bool = false;

    for &item in arr.iter() {
        if inserted == false && item > element {
            new_arr.push(element);
            inserted = true;
        }

        new_arr.push(item);
    }

    if inserted == false {
        new_arr.push(element);
    }

    new_arr
}
